crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2673,6 +2722,73 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    /// Performs lexicographic comparison between two slices of types.
+    ///
+    /// This comparison supports only equality and ordering operators: `==`, `!=`, `<`, `<=`, `>`, `>=`.
+    /// For lexicographic comparison, elements from both slices are compared pairwise using
+    /// `infer_binary_type_comparison`. If a conclusive result cannot be determined as a `BoolLiteral`,
+    /// it returns `bool`. If the comparison is undefined between types, it returns `None`.
+    fn infer_lexicographic_type_comparison(
+        &mut self,
+        left: &[Type<'db>],
+        op: ast::CmpOp,
+        right: &[Type<'db>],
+    ) -> Option<Type<'db>> {
+        if !matches!(
+            op,
+            ast::CmpOp::Eq
+                | ast::CmpOp::NotEq
+                | ast::CmpOp::Lt
+                | ast::CmpOp::LtE
+                | ast::CmpOp::Gt
+                | ast::CmpOp::GtE
+        ) {
+            return None;
+        }
```

Comment:
It looks like this is not "unsupported op between specific types" case but rather "not supported by this method" case, so I don't think we should return `None` here. It's a programming error if this method is ever called with an op not in this list, so this should just be a debug assert instead.

Reply:
I agree with your point. This is indeed a programming error, and makes meaning of `None` ambiguous.

However, as you know, lexicographic order cannot be defined outside the comparison operators `==`, `!=`, `<`, `<=`, `>`, and `>=`. Even if we use `debug_assert!`, avoiding `None` would be impossible without using `unreachable!`.

So, what if we change the type of the `op` parameter to prevent errors at compile-time instead?

In the Python docs, comparison operations are divided into three categories: [(Python docs)](https://docs.python.org/3/reference/expressions.html#value-comparisons)

1. Value Comparison
2. Membership Test Comparison
3. Identity Comparison

What do you think about defining new enums and using it like this?

```rust
enum ValueCmpOp {
    Eq,
    NotEq,
    Lt,
    LtE,
    Gt,
    GtE,
}

//...

enum CmpOpSubset {
    Value(ValueCmpOp),
    MembershipTest(MembershipTestCmpOp),
    Identity(IdentityCmpOp),
}

impl From<ast::CmpOp> for CmpOpSubset {
    // ...
}

fn infer_lexicographic_type_comparison(
    &mut self,
    left: &[Type<'db>],
    op: ValueCmpOp,
    right: &[Type<'db>],
) -> Option<Type<'db>> {
    // implementation
}
```

This would allow for an implementation like this, where only the necessary op can be passed, eliminating the need to return None in `infer_lexicographic_type_comparison`:
```rust
match op.into() {
    CmpOpSubset::Value(op) => {
        self.infer_lexicographic_type_comparison(lhs_elements, op, rhs_elements)
    }
    CmpOpSubset::MembershipTest(op) => {
        // ...
    }
    CmpOpSubset::Identity(op) => {
        // ...
    }
}
```

Reply:
I think we can merge this with the discussion in [this thread](https://github.com/astral-sh/ruff/pull/13712#discussion_r1801739049)

Reply:
I think we can merge this with the discussion in https://github.com/astral-sh/ruff/pull/13712#discussion_r1801739049